feat: add response type for feeder_gateway

chee-chyuan opened the pull request:
Fixes | Closes | Resolves #

## Description

Adding typing for feeder_gateway responses

## Changes:

changes only at `gateway.go`

## Types of changes

_Leave only items that describe your changes, remove the rest. Remove this line too._

- New feature (non-breaking change which adds functionality)

## Testing

**Requires testing**: No

**Did you write tests??**: No

## Documentation

**If this requires a documentation update, did you add one?** Yes/No

jelilat commented on clients/gateway.go:
```diff
@@ -210,9 +233,39 @@ type Abi []struct {
+// 	MainScope        interface{}   `json:"main_scope"`
+// 	Hints            interface{}   `json:"hints"`
+// 	CompilerVersion  string        `json:"compiler_version"`
+// }
```
Not sure why these are commented out.

chee-chyuan replied:
my bad. i forgot to delete it. thanks for catching this

chee-chyuan commented on clients/gateway.go:
```diff
@@ -210,9 +233,39 @@ type Abi []struct {
 		Type string `json:"type"`
 	} `json:"inputs"`
-	Outputs []interface{} `json:"outputs"`
+	Outputs []struct {
+		Name string `json:"name"`
+		Type string `json:"type"`
+	} `json:"outputs"`
+	StateMutability string `json:"stateMutability"`
+	Data            []struct {
+		Name string `json:"name"`
+		Type string `json:"type"`
+	}
+	Keys []interface{} `json:"keys"`
```
thanks. i didnt know about this

jelilat replied:
You're welcome.

We should do the same thing for the `Program` struct. Then, we can change `Hints` from `interface{}` type to `map[uint64]interface{}` type.

chee-chyuan replied:
yea you are right again! how about map[uint64][]interface{}?

jelilat replied:
Can't completely recall. But I think the last time I marshalled using `map[uint64][]interface{}`, json sometimes rearranged the array elements. But we don't want that to happen because the original order is very important when computing hashes.

chee-chyuan replied:
oh yea ok gotcha. anyway i have updated it

joshklop reviewed (changes requested):
Looks great to me as written! Can you please add unmarshal test cases for the new structs and embedded structs in `gateway_test.go.`?

chee-chyuan commented:
> 

added test cases for new struct types. however, not 100% as im ignoring alot of `interface{}` type. also added 2 tests to ensure that the fields in `Abi` and `ClassDefinition` are ordered alphabetically.  cc @jelilat 

also, made some minor changes again to the types in `gateway.go`

omerfirmak commented on clients/gateway.go:
```diff
@@ -105,12 +105,18 @@ type Transaction struct {
+	BlockNumber              *big.Int                  `json:"block_number"`
+	TransactionIndex         *big.Int                  `json:"transaction_index"`
+	Transaction              *Transaction              `json:"transaction"`
+	TransactionFailureReason *TransactionFailureReason `json:"transaction_failure_reason"`
```
Do you have an example URL that has this field in its response?

chee-chyuan replied:
yup. i only see it in the case of not enough fee at this point of time https://alpha4.starknet.io/feeder_gateway/get_transaction_receipt?transactionHash=0x2253fed5e39a8c695e282befa2e3fa67ef0dab656b71a36c32387d5c6030c4a 

i think after sierra, we will see more of this?

omerfirmak commented on clients/gateway.go:
```diff
@@ -161,27 +167,32 @@ type ExecutionResources struct {
+	L2ToL1Message            []*L2ToL1Message          `json:"l2_to_l1_messages"`
+	TransactionHash          *felt.Felt                `json:"transaction_hash"`
+	TransactionIndex         *big.Int                  `json:"transaction_index"`
+	Status                   string                    `json:"status"`
```
This and all the fields after it should go in a separate struct called `TransactionReceiptStatus`. Similar to how `Transaction` and `TransactionStatus` works.

IronGauntlets commented:
I am closing this PR since it is not applicable right now and from out perspective this should go in the rpc rather than gateway.
